systemd/watch.go

```diff
@@ -0,0 +1,155 @@
+}
+
+type Watcher struct {
+	mu        sync.RWMutex
+	clientset *kubernetes.Clientset
+	// maps are indexed by "namespace.name"
+	configs map[string][]*corev1.Pod
+	secrets map[string][]*corev1.Pod
+}
+
+func newWatcher(clientset *kubernetes.Clientset) *Watcher {
+	return &Watcher{
+		configs:   make(map[string][]*corev1.Pod),
+		secrets:   make(map[string][]*corev1.Pod),
+		clientset: clientset,
+	}
+}
+
+func (w *Watcher) Run(p Updater) error {
+	cmwatcher, err := w.clientset.CoreV1().ConfigMaps(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	swatcher, err := w.clientset.CoreV1().Secrets(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case event, ok := <-cmwatcher.ResultChan():
+			if !ok {
+				cmwatcher, _ = w.clientset.CoreV1().ConfigMaps(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+				break
+			}
+			cm, ok := event.Object.(*corev1.ConfigMap)
+			if !ok {
+				continue
+			}
+			// event.Type == watch.Deleted ??
+			if event.Type != watch.Added && event.Type != watch.Modified {
+				continue
+			}
+
+			namespace := cm.ObjectMeta.Namespace
+			name := cm.ObjectMeta.Name
+			w.mu.RLock()
+			pods := w.configs[namespace+"."+name]
+			w.mu.RUnlock()
+			if len(pods) != 0 {
+				klog.Infof("Configmap update %s/%s, notifying %s pods", namespace, name, len(pods))
+			}
+			for _, pod := range pods {
+				p.UpdateConfigMap(context.TODO(), pod, cm)
+			}
+		case event, ok := <-swatcher.ResultChan():
+			if !ok {
+				swatcher, _ = w.clientset.CoreV1().Secrets(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+				break
+			}
+			s, ok := event.Object.(*corev1.Secret)
+			if !ok {
+				continue
+			}
+			// event.Type == watch.Deleted ??
+			if event.Type != watch.Added && event.Type != watch.Modified {
+				continue
+			}
+
+			namespace := s.ObjectMeta.Namespace
+			name := s.ObjectMeta.Name
+			w.mu.RLock()
+			pods := w.secrets[namespace+"."+name]
+			w.mu.RUnlock()
+			if len(pods) != 0 {
+				klog.Infof("Secret update %s/%s, notifying %s pods", namespace, name, len(pods))
+			}
+			for _, pod := range pods {
+				p.UpdateSecret(context.TODO(), pod, s)
+			}
+		}
+	}
+}
+
+// Watch wathces the config maps and secrets for the pod. It calls back into the
+// provider using the Updater interface methods.
+func (w *Watcher) Watch(pod *corev1.Pod) {
+	if w == nil {
+		return
+	}
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	for _, v := range pod.Spec.Volumes {
+		switch {
+		case v.ConfigMap != nil:
+			key := pod.Namespace + "." + v.ConfigMap.Name
+			w.configs[key] = append(w.configs[key], pod)
+
+		case v.Secret != nil:
+			key := pod.Namespace + "." + v.Secret.SecretName
+			w.secrets[key] = append(w.secrets[key], pod)
+		}
+	}
+}
+
+// Unwatch removes the watches for the pod.
+func (w *Watcher) Unwatch(pod *corev1.Pod) {
+	if w == nil {
```

Comment:
Wait, Go allows for this to happen so it needs to be covered for? 🤯

Reply:
this can probably be removed, because the watching will not be optional, but know it depends on the getting the kube config via a flag...

yes this is valid Go :)